compiler/src/dotty/tools/dotc/reporting/WConf.scala

```diff
@@ -0,0 +1,94 @@
+package dotty.tools
+package dotc
+package reporting
+
+import dotty.tools.dotc.core.Contexts._
+
+import java.util.regex.PatternSyntaxException
+import scala.annotation.internal.sharable
+import scala.collection.mutable.ListBuffer
+import scala.util.matching.Regex
+
+enum MessageFilter:
+  def matches(message: Diagnostic): Boolean = this match {
+    case Any => true
+    case Deprecated => message.isInstanceOf[Diagnostic.DeprecationWarning]
+    case Feature => message.isInstanceOf[Diagnostic.FeatureWarning]
+    case MessagePattern(pattern) => pattern.findFirstIn(message.msg.rawMessage).nonEmpty
+  }
+  case Any, Deprecated, Feature
+  case MessagePattern(pattern: Regex)
+
+enum Action:
+  case Error, Warning, Info, Silent
+
+final case class WConf(confs: List[(List[MessageFilter], Action)]):
+  def action(message: Diagnostic): Action = confs.collectFirst {
+    case (filters, action) if filters.forall(_.matches(message)) => action
+  }.getOrElse(Action.Warning)
+
+@sharable object WConf:
```

Comment:
I don't think this is actually sharable, multiple compiler instances might call into this object at the same time leading to data races with `parsedCache`, this probably should be part of the Context instead.